### PR TITLE
Add Db2 dialect

### PR DESCRIPTION
**Pull request by ddresslerlegalplans**
### Brief summary of the change made

Fixes https://github.com/sqlfluff/sqlfluff/issues/3229

### Are there any other side effects of this change that we should be aware of?
Not that I am aware of

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.

- [x] Please confirm you have completed any of the necessary steps below.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: test/fixtures/dialects/db2fori/.sqlfluff [hunk 1]
```diff
@@ -0,0 +1,2 @@
+[sqlfluff]
+dialect = sqlite
```
Comment (tunetheweb) [inline test/fixtures/dialects/db2fori/.sqlfluff:2]: This is wrong (though not used for generation, but helpful when you parse a file manually):

```suggestion
dialect = db2fori
```

File: test/fixtures/dialects/db2fori/create_table_field_name_with_pound_sign.yml [hunk 2]
```diff
@@ -0,0 +1,49 @@
+# YML test files are auto-generated from SQL files and should not be edited by
+# hand. To help enforce this, the "hash" field in the file must match a hash
+# computed by SQLFluff when running the tests. Please run
+# `python test/generate_parse_fixture_yml.py`  to generate them after adding or
+# altering SQL files.
+
+
+# ::TODO:: run python test/generate_parse_fixture_yml.py
+# ::TODO:: figure out why when I run the command above I get the exception:
```
Comment (tunetheweb) [inline test/fixtures/dialects/db2fori/create_table_field_name_with_pound_sign.yml:9]: I don't get that exception on your branch and it looks fine to me.

Are you definitely using your local copy (e.g. `pip install -e .`)?
Or sometimes better to use tox:
```
tox -e generate-fixture-yml
```

**Review by tunetheweb — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 3]
```diff
@@ -9,6 +9,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 Note: Changes are now automatically tracked in [GitHub](https://github.com/sqlfluff/sqlfluff/releases) and will be copied in here on each release (please remember to update the issues and contributors to links!). There is no need to manually edit this file going forward.
 -->
 
+## [::TODO:: add release #] - ::TODO:: add release date
+* Add Db2 for i dialect [<::TODO::put pr # here>](https://github.com/sqlfluff/sqlfluff/pull/<::TODO::put pr # here>) [@ddresslerlegalplans](https://github.com/ddresslerlegalplans)
```
[the remark this answers is not in the document]
Comment (ddresslerlegalplans) [reply CHANGELOG.md]: Yep, no problem - https://github.com/sqlfluff/sqlfluff/pull/3231/commits/568657fd45721660b07a05530a038a8d6428f11d

File: README.md [hunk 4]
```diff
@@ -21,6 +21,7 @@ Although SQL is reasonably consistent in its implementations, there are several
 
 - ANSI SQL - this is the base version and on occasion may not strictly follow the ANSI/ISO SQL definition
 - [BigQuery](https://cloud.google.com/bigquery/)
+- [Db2 for i](https://www.ibm.com/docs/en/i/7.4?topic=overview-db2-i)
```
[the remark this answers is not in the document]
Comment (ddresslerlegalplans) [reply README.md]: I did some searching and apparently the SQL shouldn't be too different, so I'm cool with a generic Db2 to cover Db2 LUW, Db2 for i, Db2 for z/OS https://github.com/sqlfluff/sqlfluff/pull/3231/commits/74b8e243d53b70ce31aa0358cbc66960e3cc8de7  https://github.com/sqlfluff/sqlfluff/pull/3231/commits/4dbbd302bd00f5e5cf5c11c829c67a95a2e3b88e

File: test/fixtures/dialects/db2fori/.sqlfluff (hunk 1 above)
Comment (ddresslerlegalplans) [reply test/fixtures/dialects/db2fori/.sqlfluff]: https://github.com/sqlfluff/sqlfluff/pull/3231/commits/b2477587ca57b2dd106d62823827683abd93dbeb thanks!

File: test/fixtures/dialects/db2fori/create_table_field_name_with_pound_sign.yml (hunk 2 above)
Comment (ddresslerlegalplans) [reply test/fixtures/dialects/db2fori/create_table_field_name_with_pound_sign.yml]: Thanks! The tox command worked well  - https://github.com/sqlfluff/sqlfluff/pull/3231/commits/7fbacc85bc5a6a7dd7d2cf03fded6deb013b7ec5

**Review by ddresslerlegalplans — COMMENTED** [review]
Thanks for the review,  I've addressed all issues.  Let me know if I need to squash commits and rename the first commit from "Add Db2 for i dialect" to "Add Db2 dialect"

**Review by tunetheweb — CHANGES_REQUESTED** [review]
Looks pretty good. Couple of suggested changes.

Also could you update the main README.md to list this as a dialect too?

File: README.md [hunk 5]
```diff
@@ -21,6 +21,7 @@ Although SQL is reasonably consistent in its implementations, there are several
 
 - ANSI SQL - this is the base version and on occasion may not strictly follow the ANSI/ISO SQL definition
 - [BigQuery](https://cloud.google.com/bigquery/)
+- [Db2](https://en.wikipedia.org/wiki/IBM_Db2)
```
[the remark this answers is not in the document]
Comment (ddresslerlegalplans) [reply README.md:24]: This is their official site: https://www.ibm.com/analytics/db2  I can change it to that

[the remark this answers is not in the document]
Comment (ddresslerlegalplans) [reply README.md]: https://github.com/sqlfluff/sqlfluff/pull/3231/commits/4cac2d00c74fd3f325f0e8c70dc8fe7e75342d3f

File: docs/source/dialects.rst [hunk 6]
```diff
@@ -61,6 +61,15 @@ The dialect for `Google BigQuery`_.
...
 .. _`Google BigQuery`: https://cloud.google.com/bigquery/
 
+.. _db2_dialect_ref:
+
+Db2
+------
+
+The dialect for `Db2`_.
+
+.. _`Db2`: https://en.wikipedia.org/wiki/IBM_Db2
```
[the remark this answers is not in the document]
Comment (ddresslerlegalplans) [reply docs/source/dialects.rst]: https://github.com/sqlfluff/sqlfluff/pull/3231/commits/4cac2d00c74fd3f325f0e8c70dc8fe7e75342d3f

File: setup.cfg [hunk 7]
```diff
@@ -41,6 +41,7 @@ keywords =
     linter
     formatter
     bigquery
+    db2fori
```
[the remark this answers is not in the document]
Comment (ddresslerlegalplans) [reply setup.cfg]: Thanks!  Forgot to hit save. https://github.com/sqlfluff/sqlfluff/pull/3231/commits/4cac2d00c74fd3f325f0e8c70dc8fe7e75342d3f

**Review by tunetheweb — APPROVED** [review]
I think this looks good to me. You happy to merge now or still working on other changes?

**Comment by ddresslerlegalplans** [comment]
I'm happy with a merge now.  Thanks again for the help and review.

**Comment by tunetheweb** [comment]
> I'm happy with a merge now. Thanks again for the help and review.

Thank YOU for working in this!

**Comment by tunetheweb** [comment]
https://twitter.com/SQLFluff/status/1521208357176102915?s=20&t=LHzHIkUT9KW9I8F0wNUcSg

**Pull request merged**